frontend/public/components/nav/perspective-nav.tsx

```diff
@@ -38,81 +30,44 @@ const PerspectiveNav: React.FC<{}> = () => {
-    e.preventDefault();
-    e.stopPropagation();
-    confirmNavUnpinModal(resource, pinnedResources, setPinnedResources);
-  };
-
   // Until admin perspective is contributed through extensions, render static
   // `AdminNav` and any additional plugin nav items.
   if (perspective === 'admin') {
     return <AdminNav pluginNavItems={orderedNavItems} />;
   }
 
-  const getLabelForResource = (resource: string): string => {
-    const model: K8sKind | undefined = modelFor(resource);
-    if (model) {
-      if (model.labelPluralKey) {
-        return t(model.labelPluralKey);
-      }
-      return model.labelPlural || model.plural;
-    }
-    return '';
-  };
-
   const getPinnedItems = (): React.ReactElement[] =>
     pinnedResourcesLoaded
-      ? pinnedResources
-          .map((resource) => {
-            const model = modelFor(resource);
-            if (!model) {
-              return null;
-            }
-            const { apiVersion, apiGroup, namespaced, crd, plural } = model;
-            const label = getLabelForResource(resource);
-            const duplicates =
-              pinnedResources.filter((res) => getLabelForResource(res) === label).length > 1;
-            const props = {
-              key: `pinned-${resource}`,
-              name: label,
-              resource: crd ? referenceForModel(model) : plural,
-              tipText: duplicates ? `${label}: ${apiGroup || 'core'}/${apiVersion}` : null,
-              id: resource,
-            };
-            const Component: NavLinkComponent = namespaced ? ResourceNSLink : ResourceClusterLink;
-            const removeButton = (
-              <Button
-                className="oc-nav-pinned-item__unpin-button"
-                variant="link"
-                aria-label={t('public~Unpin')}
-                onClick={(e) => unPin(e, resource)}
-              >
-                <MinusCircleIcon className="oc-nav-pinned-item__icon" />
-              </Button>
-            );
-
-            return (
-              <RootNavLink
-                key={resource}
-                className="oc-nav-pinned-item"
-                component={Component}
-                {...props}
-              >
-                {removeButton}
-              </RootNavLink>
-            );
-          })
-          .filter((p) => p !== null)
+      ? (pinnedResources.length === 1
+          ? [
+              <PinnedResource
+                key="0"
+                resource={pinnedResources[0]}
+                navResources={pinnedResources}
+                onChange={setPinnedResources}
+              />,
+            ]
+          : pinnedResources.map((resource, idx) => (
+              <NavItemDnd
+                key={idx.toString()}
+                idx={idx}
+                resource={resource}
+                onChange={setPinnedResources}
+                navResources={pinnedResources}
+              />
+            ))
+        ).filter((p) => p !== null)
```

Comment:
If I'm understanding correctly, this filter is used to remove invalid resource types from the list, which is done with the `modelFor` check in `<PinnedResource>`.

If so, can we do the filtering before creating nav components?
Because if `pinnedResources` contains invalid resource types, and the length is greater than one, then the drag icon would be shown even if there is only one nav item after filtering.

![0](https://user-images.githubusercontent.com/20013884/157099436-44d8c4c8-f123-4db4-b50f-891e6cc5a32e.png)

Reply:
Thanks for catching this. Fixed it and also added a test. PTAL